File: python/src/resolve/resolve.py
from typing import Union, Literal, TypedDict, List, Optional

from solana.rpc.async_api import AsyncClient
from solders.account import Account
from solders.pubkey import Pubkey
from typing_extensions import NotRequired

from NameRegistryState import NameRegistryState
from constants import SIGNATURE_LENGTH_IN_BYTES
from custom_types import Record
from exception import (
    DomainDoesNotExistException,
    CouldNotFindNftOwnerException,
    NoAccountDataException,
    PdaOwnerNotAllowedException, WrongValidationException, RecordMalformedException, InvalidRoAException,
)
from nft import retrieve_nft_owner
from nft.constants import NAME_TOKENIZER_ID
from nft.state import NftRecord, Tag
from record import get_record_key, check_sol_record
from record_v2.get_record_key_v2 import get_record_key_v2
from utils import get_domain_key
from sns_records.state import Record as RecordV2, Validation

AllowPda = Union[Literal["any"], bool]


class ResolveConfig(TypedDict):
    allow_pda: AllowPda
    program_ids: NotRequired[List[Pubkey]]


def _check_sol_record_v2(sol_record_v2_info: Account, registry: NameRegistryState):
    record_v2 = RecordV2.deserialize(sol_record_v2_info.data)
    staleness_id = record_v2.get_staleness_id()
    roa_id = record_v2.get_roa_id()
    content = record_v2.get_content()
    print(record_v2.header.right_of_association_validation)
    print(record_v2.header.staleness_validation)
    print(Validation.Solana.value)
    if len(content) != 32:
        raise RecordMalformedException("Record is malformed")

    if (record_v2.header.right_of_association_validation != Validation.Solana.value or
            record_v2.header.staleness_validation != Validation.Solana.value):
        raise WrongValidationException()

    if not staleness_id == bytes(registry.owner):
        return None

    if roa_id == content:
        return Pubkey(content)

    raise InvalidRoAException(
        f"The RoA ID should be {str(Pubkey(content))} "
        f"but is {str(Pubkey(roa_id))}"
    )



async def resolve(
    connection: AsyncClient, domain: str, config: Optional[ResolveConfig] = None
) -> Pubkey:
    if config is None:
        config = {"allow_pda": False}

    pubkey = get_domain_key(domain)["pubkey"]
    nft_record_key, _ = await NftRecord.find_key(pubkey, NAME_TOKENIZER_ID)
    sol_record_key_v1 = get_record_key(domain, Record.SOL)
    sol_record_key_v2 = get_record_key_v2(domain, Record.SOL)
    res = await connection.get_multiple_accounts(
        [nft_record_key, sol_record_key_v1, sol_record_key_v2, pubkey]
    )
    nft_record_info = res.value[0]
    sol_record_info_v1 = res.value[1]
    sol_record_info_v2 = res.value[2]
    registry_info = res.value[3]

    if not registry_info.data:
        raise DomainDoesNotExistException(f"Domain does not exist: {domain}")

    registry = NameRegistryState.deserialize(registry_info.data)

    if nft_record_info and nft_record_info.data:
        nft_record = NftRecord.deserialize(nft_record_info.data)
        if nft_record.tag == Tag.active_record.value:
            nft_owner = await retrieve_nft_owner(connection, pubkey)
            if nft_owner is None:
                raise CouldNotFindNftOwnerException(
                    f"Could not find NFT owner for domain: {domain}"
                )
            return nft_owner

    if sol_record_info_v2 and sol_record_info_v2.data:
        resolved = _check_sol_record_v2(sol_record_info_v2, registry)
        if resolved:
            return resolved

    # Check SOL record V1
    if sol_record_info_v1 and sol_record_info_v1.data:
        expected_buffer = sol_record_info_v1.data[
            NameRegistryState.HEADER_LEN : NameRegistryState.HEADER_LEN + 32
        ] + bytes(sol_record_key_v1)

        expected = expected_buffer.hex().encode()
        valid = check_sol_record(
            expected,
            sol_record_info_v1.data[
                NameRegistryState.HEADER_LEN + 32 : NameRegistryState.HEADER_LEN
                + 32
                + SIGNATURE_LENGTH_IN_BYTES
            ],
            registry.owner,
        )
        if valid:
            return Pubkey(
                sol_record_info_v1.data[
                    NameRegistryState.HEADER_LEN : NameRegistryState.HEADER_LEN + 32
                ]
            )

    # Check if the registry owner is a PDA
    is_on_curve = Pubkey.is_on_curve(registry.owner)
    if not is_on_curve:
        if config["allow_pda"] == "any":
            return registry.owner
        elif config["allow_pda"]:
            owner_info = await connection.get_account_info(registry.owner)
            if owner_info.value is None:
                raise NoAccountDataException(
                    f"No account data for PDA: {registry.owner}"
                )
            is_allowed = any(
                e == owner_info.value.owner for e in config.get("program_ids", [])
            )
            if is_allowed:
                return registry.owner
            raise PdaOwnerNotAllowedException(
                f"PDA owner not allowed: {owner_info.value.owner}"
            )
        else:
            raise PdaOwnerNotAllowedException()

    return registry.owner

File: js/src/resolve/resolve.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolve = void 0;
const web3_js_1 = require("@solana/web3.js");
const getDomainKeySync_1 = require("../utils/getDomainKeySync");
const state_1 = require("../nft/state");
const const_1 = require("../nft/const");
const getRecordKeySync_1 = require("../record/getRecordKeySync");
const record_1 = require("../types/record");
const getRecordV2Key_1 = require("../record_v2/getRecordV2Key");
const sns_records_1 = require("@bonfida/sns-records");
const error_1 = require("../error");
const state_2 = require("../state");
const checkSolRecord_1 = require("../record/checkSolRecord");
const retrieveNftOwnerV2_1 = require("../nft/retrieveNftOwnerV2");
/**
 * Resolve function according to SNS-IP 5
 * @param connection
 * @param domain
 * @param config
 * @returns
 */
const resolve = (connection_1, domain_1, ...args_1) => __awaiter(void 0, [connection_1, domain_1, ...args_1], void 0, function* (connection, domain, config = { allowPda: false }) {
    var _a;
    const { pubkey } = (0, getDomainKeySync_1.getDomainKeySync)(domain);
    const [nftRecordKey] = state_1.NftRecord.findKeySync(pubkey, const_1.NAME_TOKENIZER_ID);
    const solRecordV1Key = (0, getRecordKeySync_1.getRecordKeySync)(domain, record_1.Record.SOL);
    const solRecordV2Key = (0, getRecordV2Key_1.getRecordV2Key)(domain, record_1.Record.SOL);
    const [nftRecordInfo, solRecordV1Info, solRecordV2Info, registryInfo] = yield connection.getMultipleAccountsInfo([
        nftRecordKey,
        solRecordV1Key,
        solRecordV2Key,
        pubkey,
    ]);
    if (!(registryInfo === null || registryInfo === void 0 ? void 0 : registryInfo.data)) {
        throw new error_1.DomainDoesNotExist(`Domain ${domain} does not exist`);
    }
    const registry = state_2.NameRegistryState.deserialize(registryInfo.data);
    // If NFT record active -> NFT owner is the owner
    if (nftRecordInfo === null || nftRecordInfo === void 0 ? void 0 : nftRecordInfo.data) {
        const nftRecord = state_1.NftRecord.deserialize(nftRecordInfo.data);
        if (nftRecord.tag === state_1.Tag.ActiveRecord) {
            const nftOwner = yield (0, retrieveNftOwnerV2_1.retrieveNftOwnerV2)(connection, pubkey);
            if (!nftOwner) {
                throw new error_1.CouldNotFindNftOwner();
            }
            return nftOwner;
        }
    }
    // Check SOL record V2
    recordV2: if (solRecordV2Info === null || solRecordV2Info === void 0 ? void 0 : solRecordV2Info.data) {
        const recordV2 = sns_records_1.Record.deserialize(solRecordV2Info.data);
        const stalenessId = recordV2.getStalenessId();
        const roaId = recordV2.getRoAId();
        const content = recordV2.getContent();
        if (content.length !== 32) {
            throw new error_1.RecordMalformed(`Record is malformed`);
        }
        console.log(recordV2.header.rightOfAssociationValidation);
        console.log(recordV2.header.stalenessValidation);
        console.log(sns_records_1.Validation.Solana);
        if (recordV2.header.rightOfAssociationValidation !== sns_records_1.Validation.Solana ||
            recordV2.header.stalenessValidation !== sns_records_1.Validation.Solana) {
            console.log("oui");
            throw new error_1.WrongValidation();
        }
        if (!stalenessId.equals(registry.owner.toBuffer())) {
            break recordV2;
        }
        if (roaId.equals(content)) {
            return new web3_js_1.PublicKey(content);
        }
        console.log("non");
        throw new error_1.InvalidRoAError(`The RoA ID shoudl be ${new web3_js_1.PublicKey(content).toBase58()} but is ${new web3_js_1.PublicKey(roaId).toBase58()} `);
    }
    // Check SOL record V1
    if (solRecordV1Info === null || solRecordV1Info === void 0 ? void 0 : solRecordV1Info.data) {
        const encoder = new TextEncoder();
        const expectedBuffer = Buffer.concat([
            solRecordV1Info.data.slice(state_2.NameRegistryState.HEADER_LEN, state_2.NameRegistryState.HEADER_LEN + 32),
            solRecordV1Key.toBuffer(),
        ]);
        const expected = encoder.encode(expectedBuffer.toString("hex"));
        const valid = (0, checkSolRecord_1.checkSolRecord)(expected, solRecordV1Info.data.slice(state_2.NameRegistryState.HEADER_LEN + 32, state_2.NameRegistryState.HEADER_LEN + 32 + web3_js_1.SIGNATURE_LENGTH_IN_BYTES), registry.owner);
        if (valid) {
            return new web3_js_1.PublicKey(solRecordV1Info.data.slice(state_2.NameRegistryState.HEADER_LEN, state_2.NameRegistryState.HEADER_LEN + 32));
        }
    }
    // Check if the registry owner is a PDA
    const isOnCurve = web3_js_1.PublicKey.isOnCurve(registry.owner);
    if (!isOnCurve) {
        if (config.allowPda === "any") {
            return registry.owner;
        }
        else if (config.allowPda) {
            const ownerInfo = yield connection.getAccountInfo(registry.owner);
            const isAllowed = (_a = config.programIds) === null || _a === void 0 ? void 0 : _a.some((e) => { var _a; return (_a = ownerInfo === null || ownerInfo === void 0 ? void 0 : ownerInfo.owner) === null || _a === void 0 ? void 0 : _a.equals(e); });
            if (isAllowed) {
                return registry.owner;
            }
            throw new error_1.PdaOwnerNotAllowed(`The Program ${ownerInfo === null || ownerInfo === void 0 ? void 0 : ownerInfo.owner.toBase58()} is not allowed`);
        }
        else {
            throw new error_1.PdaOwnerNotAllowed();
        }
    }
    return registry.owner;
});
exports.resolve = resolve;
